drivers/regulator/regulator_cp9314.c

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright (c) 2024 Cirrus Logic, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cirrus_cp9314
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/sys/util.h>
+
+#include "regulator_cp9314.h"
+
+enum cp9314_sync_roles {
+	CP9314_ROLE_HOST,
+	CP9314_ROLE_DEV2,
+	CP9314_ROLE_DEV3,
+	CP9314_ROLE_DEV4,
+	CP9314_ROLE_STANDALONE,
+};
+
+struct regulator_cp9314_config {
+	struct regulator_common_config common;
+	struct i2c_dt_spec i2c;
+	struct gpio_dt_spec en_pin;
+};
+
+struct regulator_cp9314_data {
+	struct regulator_common_data data;
+	enum cp9314_sync_roles sync_role;
+	uint8_t rev_id;
+	uint8_t op_mode;
+};
+
+struct cp9314_reg_patch {
+	uint8_t reg_addr;
+	uint8_t mask;
+	uint8_t value;
+};
+
+static struct cp9314_reg_patch b0_reg_patch[18] = {
+	{CP9314_REG_CRUS_CTRL, GENMASK(7, 0), CP9314_CRUS_KEY_UNLOCK},
+	{CP9314_REG_LION_COMP_CTRL_3, CP9314_VIN_OV_CFG, 0x1B},
+	{CP9314_REG_LION_COMP_CTRL_1, CP9314_VOUT_OV_CFG_0, 0x30},
+	{CP9314_REG_LION_COMP_CTRL_2, CP9314_VOUT_OV_CFG_1, 0xC},
+	{CP9314_REG_VIN2OUT_OVP, CP9314_VIN2OUT_OVP, 0x2},
+	{CP9314_REG_VIN2OUT_UVP, CP9314_VIN2OUT_UVP, 0x1},
+	{CP9314_REG_VOUT_UVP, CP9314_VOUT_UVP_DIS, 0},
+	{CP9314_REG_VOUT_UVP, CP9314_VOUT_UVP, 0},
+	{CP9314_REG_LION_COMP_CTRL_1, CP9314_VIN_SWITCH_OK_DIS_0, 0},
+	{CP9314_REG_LION_COMP_CTRL_4, CP9314_VIN_SWITCH_OK_DIS_1, 0},
+	{CP9314_REG_LION_COMP_CTRL_1, CP9314_VIN_SWITCH_OK_CFG, 0},
+	{CP9314_REG_LION_CFG_3, CP9314_LB_MIN_FREQ_SEL_0, 0x80},
+	{CP9314_REG_LB_CTRL, CP9314_LB_MIN_FREQ_SEL_1, 0x4},
+	{CP9314_REG_TRIM_8, CP9314_MODE_CTRL_UPDATE_BW_0, 0x2},
+	{CP9314_REG_LION_CFG_3, CP9314_MODE_CTRL_UPDATE_BW_1, 0x2},
+	{CP9314_REG_IIN_OCP, CP9314_IIN_OCP_DIS, CP9314_IIN_OCP_DIS},
+	{CP9314_REG_IIN_PEAK_OCP, CP9314_IIN_PEAK_OCP_DIS, CP9314_IIN_PEAK_OCP_DIS},
+	{CP9314_REG_CRUS_CTRL, GENMASK(7, 0), CP9314_CRUS_KEY_LOCK},
+};
+
+static struct cp9314_reg_patch otp_1_patch[3] = {
+	{CP9314_REG_OPTION_REG_1, CP9314_LB1_DELAY_CFG, 0},
+	{CP9314_REG_BST_CP_PD_CFG, CP9314_LB1_BLANK_CFG, CP9314_LB1_BLANK_CFG},
+	{CP9314_REG_TSBAT_CTRL, CP9314_LB1_STOP_PHASE_SEL, CP9314_LB1_STOP_PHASE_SEL},
+};
```

Comment:
ping

Reply:
ack, let me know if the added comments are sufficient.

Reply:
It could do with an overall comment for the block giving a rough indication of what it is for, so that users can see what is being fixed.
e.g. 
`/** Patch for b0 silicon to correct trim errata and feedback scaling errors */`